crates/flux-syntax/src/parser/mod.rs

```diff
@@ -630,8 +630,47 @@ fn parse_indices(cx: &mut ParseCtxt) -> ParseResult<Indices> {
     Ok(Indices { indices, span: cx.mk_span(lo, hi) })
 }
 
+fn parse_fn_bound_input(cx: &mut ParseCtxt) -> ParseResult<GenericArg> {
+    let lo = cx.lo();
+    let tys = parens(cx, Comma, parse_type)?;
+    let hi = cx.hi();
+    let kind = TyKind::Tuple(tys);
+    let span = cx.mk_span(lo, hi);
+    let in_ty = Ty { kind, node_id: cx.next_node_id(), span };
+    Ok(GenericArg { kind: GenericArgKind::Type(in_ty), node_id: cx.next_node_id() })
+}
+
+fn parse_fn_bound_output(cx: &mut ParseCtxt) -> ParseResult<GenericArg> {
+    let lo = cx.lo();
+
+    let ty = if cx.advance_if(Tok::RArrow) {
+        parse_type(cx)?
+    } else {
+        Ty { kind: TyKind::Tuple(vec![]), node_id: cx.next_node_id(), span: cx.mk_span(lo, lo) }
+    };
+    let hi = cx.hi(); // should this be `hi` or `lo`?
+    let ident = Ident { name: Output, span: cx.mk_span(lo, hi) };
+    Ok(GenericArg { kind: GenericArgKind::Constraint(ident, ty), node_id: cx.next_node_id() })
+}
+
+fn parse_fn_bound_path(cx: &mut ParseCtxt) -> ParseResult<Path> {
+    let lo = cx.lo();
+    let ident = parse_ident(cx)?;
+    let in_arg = parse_fn_bound_input(cx)?;
+    let out_arg = parse_fn_bound_output(cx)?;
+    let args = vec![in_arg, out_arg]; // inputs and output
+    let segment = PathSegment { ident, args, node_id: cx.next_node_id() };
+    let hi = cx.hi();
+    Ok(Path {
+        segments: vec![segment],
+        refine: vec![],
+        node_id: cx.next_node_id(),
+        span: cx.mk_span(lo, hi),
+    })
+}
```

Comment:
Maybe in another PR, but I think it would make things easier if the AST is closer to the syntax, like in rustc, where they separate between parenthesized args and angle-bracketed args (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/enum.GenericArgs.html). In particular, I think it would help with the "scopes" in `refinement_resolver.rs`

Reply:
Ok I'll pull into a separate issue.